server/channels/web/handlers.go

```diff
@@ -401,7 +401,15 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			} else {
 				endpoint = h.HandlerName
 			}
-			c.App.Metrics().ObserveAPIEndpointDuration(endpoint, r.Method, statusCode, elapsed)
+			originDevice := "unknown"
+			if session := c.AppContext.Session(); session != nil {
+				originDevice = "web"
+				if session.IsMobileApp() {
+					originDevice = "mobile"
+				}
+			}
```

Comment:
I wonder if we could rely on the request's user agent here to differentiate between the three platforms (web,desktop,mobile) when missing any better euristic. From what I remember mobile adds some special prefixes (e.g. `rn` and `Mattermost`). @cpoile Investigated some of this a while back, may be worth checking with him.

Desktop should also have a special `Mattermost` suffix.

Reply:
Hey @agarciamontoro, like Claudio said, Desktop will always have a `Mattermost` in the user agent (also "electron", but `Mattermost` will be good enough). 
As for mobile, do you need to know if it's v2 or not?  Here's our full test for post-GA v2:

```go
func isMobilePostGA(r *http.Request) (mobile, postGA bool) {
	queryParam := r.URL.Query().Get("mobilev2")
	if queryParam == "true" {
		return true, true
	}

	// Below here is to test two things: is this mobile pre-GA? Is mobile version 441
	// (a one-week period when we didn't have the above queryParam)
	// TODO: simplify this once we can stop supporting 441.
	//   https://mattermost.atlassian.net/browse/MM-48929
	userAgent := r.Header.Get("User-Agent")
	fields := strings.Fields(userAgent)
	clientAgent := fields[0] // safe to assume there's at least one
	isMobile := strings.HasPrefix(clientAgent, "rnbeta") || strings.HasPrefix(clientAgent, "Mattermost")
	if !isMobile {
		return false, false
	}
	agent := strings.Split(clientAgent, "/")
	if len(agent) != 2 {
		return true, false
	}

	// We can't use a semver package, because we're not using semver correctly. So manually parse...
	version := strings.Split(agent[1], ".")
	if len(version) != 4 {
		return true, false
	}
	minor, err := strconv.Atoi(version[3])
	if err != nil {
		return true, false
	}
	return true, minor >= 441
}

```

So if it's any mobile version, checking for `Mattermost` prefix is good enough. Hope that helps?

Reply:
Oooh, this is good, thank you! I have one more question, though: you said that Desktop can be identified by having `Mattermost` in the user agent, but that's also what identifies mobile? I guess I'm missing something.

Reply:
np: Desktop will have `Mattermost/version#` as the last field, while Mobile will have `Mattermost` as the first field.

Reply:
So intuitive! :joy: Thanks for sharing this obscure knowledge, I feel you must have suffered a bit to get it yourself.

Reply:
Hah, just wait for when it gets changed. :)